crates/recursor/src/lib.rs

```diff
@@ -35,3 +35,7 @@ pub use hickory_proto as proto;
 pub use hickory_resolver as resolver;
 pub use hickory_resolver::config::NameServerConfig;
 pub use recursor::Recursor;
+
+fn is_security_aware() -> bool {
+    cfg!(feature = "dnssec")
```

Comment:
question: do we want to be able to make the recursive resolver "security unaware" via the configuration file even when it has been compiled with DNSSEC support? if yes, then this would need to read from the configuration file or `StoreConfig` would need to store such information.

Note that "security aware" does not equal to "*validating* security resolver" (whose configuration is being discussed in RFC #2194). A non-validating, security-aware resolver can still be used with `delv`, which does not rely on the resolver doing the validation but instead does the validation on the client (`delv`) side.

AFAICT, `unbound` and BIND are always security-aware, even when DNSSEC validation is disabled.

Reply:
This goes back to an early decision in the project to split DNSSEC records out from the "base" set of DNS records. There was a concern from someone who was more deeply involved in the project some years ago that the code was easier to audit by removing a lot of the DNSSEC records from the scope of those code audits. That is, they didn't need the extra DNSSEC overhead, so we feature flagged it off.

I think it's worth us discussing how we might want to handle that. As it stands, when the `dnssec` feature is off, the records are treated as unknown types. Which means that no special logic should be triggered when coming across those record types in responses or requests, that is, they'd be treated pretty much like a normal record, similar to a `TXT` record. I'm open to us revisiting this, and how we want to treat these things, but hopefully the background here is helpful.

Reply:
thanks for the context. I think it makes sense to be able to disable "parsing" / special treatment of DNSSEC records at compile time. it reduces the binary-level attack surface, e.g. via stack smashing. it probably also has a small effect in reducing the binary size. 

I think I ought to make my question more specific: if the `dnssec` feature is enabled, do we want to provide a configuration option (in `named.toml`) to make the `hickory-dns` binary non-security-aware?

at first glance, it seems redundant since one can build from source with the `dnssec` feature disabled and achieve the same (or a better) effect. however, thinking long term, if / when `hickory-dns` is distributed in binary formats in Linux distributions then I think it would make sense to provide an opt-out security-aware configuration setting as most people would prefer the binary package (built with all features enabled) over building from source. 

disabling security-awareness via a configuration setting does not reduce the binary-level attack surface because all the DNSSEC handling logic / branches will remain in the binary, but people that want a security-hardened binary can build from source with unused features, like `dnssec`, disabled. (though, I don't see disabling `dnssec` being too congruent with security hardening)

in any case, adding an opt-out security-aware configuration option is a non-breaking change so it can be done at any time -- a decision does not have to be taken here. 

I think the `dnssec` feature should remain in place. disabling it makes `hickory-dns` non-security-aware; enabling it, makes it security-aware. that seems like an appropriate / sufficient configuration option while the main way to get `hickory-dns` is to build it yourself from source.

Reply:
I don't think we should make the security awareness depend only on the Cargo feature. As always, Cargo features can be enabled implicitly by dependencies and given the nature of DNSSEC (which has often inadvertently caused breaking stuff) IMO we should "fail open" here in the case of users only enabling the Cargo feature, and require that security awareness is turned on only if this has been explicitly wired up through API (the availability of which should probably depend on the Cargo feature).

Reply:
@djc that's a fair point. regarding the configurability you are suggesting:

In the `hickory-recursor` library the `Recursor` constructor looks like this:

``` rust
impl Recursor {
    pub fn new(
        roots: impl Into<NameServerConfigGroup>,
        ns_cache_size: usize,
        record_cache_size: usize,
    ) -> Result<Self, ResolveError> {
```

to gate the security-aware setting behind the Cargo feature we could add a method to the `Recursor`

``` rust
impl Recursor {
    /// Makes the recursive recursor security aware
    #[cfg(feature = "dnssec")]
    pub fn security_aware(&mut self) { /* .. */ }
}
```

Not sure if we want allow toggling that setting (at runtime) in between calls to `resolve`

Another option is to put all configuration settings behind a "builder" that replaces the `Recursor::new` constructor. Perhaps something like this:

``` rust
struct RecursorBuilder {
    ns_cache_size: usize,
    record_cache_size: usize,
    #[cfg(feature = "dnssec")]
    security_aware: bool,
}

impl Default for RecursorBuilder {
    fn default() -> Self {
        Self {
            // ..
            #[cfg(feature = "dnssec")]
            security_aware: false,
        }
    }
}

impl RecursorBuilder {
    /// Enable or disable security awareness
    #[cfg(feature = "dnssec")]
    pub fn security_aware(&mut self, on: bool) {
        self.security_aware = on;
    }

    pub fn build(&self) -> Result<Self, ResolveError> {
        // .. similar to what `Recursor::new` does today
    }
}
```

I like this "builder pattern" option better but it involves breaking changes. `hickory-recursor`'s Cargo.toml metadata says that the library is "experimental" so perhaps these breaking changes in the API are fine? as in it's ok to make semver-breaking releases of `hickory-recursor` at a faster pace than `hickory-dns`

---

then at the `hickory-dns` binary level the security-aware setting should be exposed in the `named.toml` file. I agree that trying that enable that setting in `named.toml` when the binary was built without the "dns" feature should result in an error. I think `#[serde(deny_unknown_fields)]` can take care of that; otherwise we can implement post-parse validation to reject such configuration.

Reply:
Yup, builder pattern sounds good to me! We generally have fairly frequent breaking changes and this is definitely okay for the recursor specifically.

Reply:
Yes. I’ve been wanting to migrate to builders for a while. This looks like a good breaking change as they go.